hikari/impl/buckets.py

```diff
@@ -281,12 +276,19 @@ def update_rate_limit(self, remaining: int, limit: int, reset_at: float) -> None
             The total calls allowed in this time window.
         reset_at : builtins.float
             The epoch at which to reset the limit.
+        force : builtins.bool
+            Whether to force updating parameters, ignoring checks.
 
         !!! note
             The `reset_at` epoch is expected to be a `time.monotonic_timestamp`
             monotonic epoch, rather than a `time.time` date-based epoch.
         """
-        self.remaining = remaining
+        # Only set remaining if its less than what we have stored internally.
+        # This is done because this function may be called with outdated values
```

Comment:
Won't this mess with the handling of endpoints which have multiple buckets also how will the stats be stored for an endpoint which just went past it's reset time and therefore self.remaining is less than the actual remaining. Also isn't this just as botched equivalent to locking per bucket?

Reply:
I forgot to account for those cases :sweat_smile:. I initially tried to implement per bucket locks, but there is really no "nice" way to do so. I could give it another try and see what I can do

Reply:
> Won't this mess with the handling of endpoints which have multiple buckets

Those are a weird mess which we can't really account for in any way.

> also how will the stats be stored for an endpoint which just went past it's reset time and therefore self.remaining is less than the actual remaining

Endpoints are updated internally when calling `is_ratelimited`, which is done when calling `acquire` on non `UNKNOWN` buckets.

Reply:
Changed to per-bucket lock :+1: